File: Design-Patterns-Using-JavaScript/Momento/MomentoPatternJson.js
class Memento {
    constructor(state) {
        this.state = state;
    }

    getState() {
        return this.state;
    }
}

class Originator {
    setState(state) {
        this.state = state;
    }

    getState() {
        return { state: this.state };
    }

    createMemento() {
        return new Memento({ state: this.state });
    }

    setMemento(m) {
        console.log(m.state);
        this.state = m.state;
    }
}

class CareTaker {
    constructor() {
        this.history = [];
        this.top = -1;
        this.max = -1;
    }

    addMemento(m) {
        this.top += 1;
        this.max = this.top;
        if (this.top <= this.history.length - 1) {
            this.history[this.top] = JSON.parse(JSON.stringify(m));
        } else {
            this.history.push(JSON.parse(JSON.stringify(m)));
        }
    }

    getMemento(index) {
        return this.history[index];
    }

    undo() {
        console.log("Undoing state.");
        if (this.top <= 0) {
            return JSON.parse(JSON.stringify(this.getMemento(0)));
        }

        this.top -= 1;
        return JSON.parse(JSON.stringify(this.getMemento(this.top)));
    }

    redo() {
        console.log("Redoing state.");
        if (this.top >= this.history.length - 1 || this.top >= this.max) {
            return JSON.parse(JSON.stringify(this.getMemento(this.top)));
        }

        this.top += 1;
        return JSON.parse(JSON.stringify(this.getMemento(this.top)));
    }

    getStatesCount() {
        return this.history.length;
    }
}

// Client code
const originator = new Originator();
const careTaker = new CareTaker();

originator.setState("State 1");
careTaker.addMemento(originator.createMemento());
console.log(originator.getState());

originator.setState("State 2");
careTaker.addMemento(originator.createMemento());
console.log(originator.getState());

originator.setState("State 3");
careTaker.addMemento(originator.createMemento());
console.log(originator.getState());

originator.setMemento(careTaker.undo());
console.log(originator.getState());

originator.setMemento(careTaker.undo());
console.log(originator.getState());

originator.setState("State 4");
careTaker.addMemento(originator.createMemento());
console.log(originator.getState());

originator.setMemento(careTaker.redo());
console.log(originator.getState());

originator.setMemento(careTaker.redo());
console.log(originator.getState());

originator.setMemento(careTaker.redo());
console.log(originator.getState());

/*
{ state: 'State 1' }
{ state: 'State 2' }
{ state: 'State 3' }
Undoing state.
{ state: 'State 2' }
{ state: { state: 'State 2' } }
Undoing state.
{ state: 'State 1' }
{ state: { state: 'State 1' } }
{ state: 'State 4' }
Redoing state.
{ state: 'State 4' }
{ state: { state: 'State 4' } }
Redoing state.
{ state: 'State 4' }
{ state: { state: 'State 4' } }
Redoing state.
{ state: 'State 4' }
{ state: { state: 'State 4' } }
*/

File: Design-Patterns-Using-Ruby/Momento/MomentoPatternJson.rb
require 'json'

class Memento
    attr_reader :state

    def initialize(state)
        @state = state
    end

    def clone
        Memento.new(JSON.parse(@state.to_json))
    end
end

class Originator
    attr_accessor :state

    def set_state(state)
        @state = state
    end

    def get_state
        @state
    end

    def create_memento
        Memento.new(JSON.parse(@state.to_json))
    end

    def set_memento(memento)
        @state = JSON.parse(memento.state.to_json)
    end
end

class CareTaker
    attr_reader :history, :top, :max

    def initialize
        @history = []
        @top = -1
        @max = -1
    end

    def add_memento(memento)
        @top += 1
        @max = @top

        if @top <= @history.size - 1
            @history[@top] = memento.clone
        else
            @history << memento.clone
        end
    end

    def get_memento(index)
        @history[index]
    end

    def undo
        puts 'Undoing state.'
        if @top <= 0
            @top = 0
            return get_memento(0)
        end

        @top -= 1
        get_memento(@top)
    end

    def redo
        puts 'Redoing state.'
        if @top >= (@history.size - 1) || @top >= @max
            return get_memento(@top)
        end

        @top += 1
        get_memento(@top)
    end

    def states_count
        @history.size
    end
end

# Client code
originator = Originator.new
care_taker = CareTaker.new

originator.set_state('State 1')
care_taker.add_memento(originator.create_memento)
puts originator.get_state

originator.set_state('State 2')
care_taker.add_memento(originator.create_memento)
puts originator.get_state

originator.set_state('State 3')
care_taker.add_memento(originator.create_memento)
puts originator.get_state

originator.set_memento(care_taker.undo)
puts originator.get_state

originator.set_memento(care_taker.undo)
puts originator.get_state

originator.set_state('State 4')
care_taker.add_memento(originator.create_memento)
puts originator.get_state

originator.set_memento(care_taker.redo)
puts originator.get_state

originator.set_memento(care_taker.redo)
puts originator.get_state

originator.set_memento(care_taker.redo)
puts originator.get_state
